snapcraft/internal/project_loader/_extensions/opengl.py

```diff
@@ -0,0 +1,100 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018-2020 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Import types and tell flake8 to ignore the "unused" List.
+
+import logging
+from typing import Any, Dict, Tuple
+
+from ._extension import Extension
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExtensionImpl(Extension):
+    """This extension eases creation of snaps that integrate with OpenGL.
+
+    For easier desktop integration, it also configures each application
+    entry with these additional plugs if strictly confined (or devmode):
+
+    \b
+    - opengl (https://snapcraft.io/docs/opengl-interface)
+    """
+
+    @staticmethod
+    def get_supported_bases() -> Tuple[str, ...]:
+        return ("core18",)
+
+    @staticmethod
+    def get_supported_confinement() -> Tuple[str, ...]:
+        return ("strict", "devmode", "classic")
+
+    def __init__(self, *, extension_name: str, yaml_data: Dict[str, Any]) -> None:
+        super().__init__(extension_name=extension_name, yaml_data=yaml_data)
+
+        logger.warning(
+            "*EXPERIMENTAL*: The OpenGL extension is experimental and likely to change."
+        )
+
+        if yaml_data.get("confinement") == "classic":
+            # Use environment variables for classic snaps.
+            self.root_snippet = {
+                "environment": {
+                    "__EGL_VENDOR_LIBRARY_DIRS": "$SNAP/etc/glvnd/egl_vendor.d:$SNAP/usr/share/glvnd/egl_vendor.d",
+                    "LIBGL_DRIVERS_PATH": "$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/dri",
+                }
+            }
+
+            # Mesa must use no-patchelf, otherwise driver will crash.
+            # TODO: develop/upstream patch to fix mesa.
+            self.parts = {
```

Comment:
Why only adding this in classic confinement?

Reply:
In this case, mesa is being added in this way only to ensure `no-patchelf` attribute is set for classic.  AFAIK, it's only a problem for classic snaps..?

Reply:
On second thought, this could (should?) be shared for both.   no-patchelf is basically a no-op in the strict case, and the `prime` cleanup would be nice to have.